### ISSUE TITLE
Implement server side search

### ISSUE THREAD

**Issue by kegsay** [matrix-org/dendrite#1307]
Spec: https://matrix.org/docs/spec/client_server/r0.6.1#id104
Sytests:
```
    × Can search for an event by body
    × Can get context around search results
    × Can back-paginate search results
    × Search works across an upgraded room and its predecessor
    × Search results with rank ordering do not include redacted events
    × Search results with recent ordering do not include redacted events
```

**Comment by S7evinK** [comment]
Implemented in #2675 and #2480

**Comment by aaronraimist** [comment]
@S7evinK the README needs updating, it says "although there are still some missing features (like Search)"
